src/librbd/AioObjectRequest.cc

```diff
@@ -623,6 +624,13 @@ void AioObjectTruncate::send_write() {
   }
 }
 
+void AioObjectTrim::pre_object_map_update(uint8_t *new_state) {
+  *new_state = OBJECT_PENDING;
+
+  // object map state (before pre) decides if a post updation is needed
+  need_post = m_ictx->object_map->update_required(m_object_no, *new_state);
```

Comment:
Before `send_pre()` is called (for each object), the object state for the range `[start, end)` has already been moved to `OBJECT_PENDING`. Therefore, pre operation for every object being trimmed becomes a no-op. But in `send_post()`, the object map state for every object is still in `OBJECT_PENDING` state as bulk object map update of `[start, end)` is done *after* objects in this range are trimmed (in `TrimRequest::send_post_copyup()`). This post object map update is done for each object being trimmed.

This is the reason to introduce `need_post` in `AioObjectTrim` -- post update is done only if the object state was not already in `OBJECT_PENDING` state.